contracts/manifold/burnredeem/ERC721BurnRedeem.sol

```diff
@@ -0,0 +1,330 @@
+
+        // Check that each group in the burn set is satisfied
+        uint256[] memory groupCounts = new uint256[](_burnRedeem.burnSet.length);
+
+        for (uint256 i = 0; i < burnTokens.length; i++) {
+            BurnToken calldata burnToken = burnTokens[i];
+            BurnItem memory burnItem = _burnRedeem.burnSet[burnToken.groupIndex].items[burnToken.itemIndex];
+
+            _validateBurnItem(burnItem, burnToken.contractAddress, burnToken.id, burnToken.merkleProof);
+            _burn(msg.sender, burnToken.contractAddress, burnToken.id);
+
+            groupCounts[burnToken.groupIndex] += 1;
+        }
+
+        for (uint256 i = 0; i < groupCounts.length; i++) {
+            require(groupCounts[i] == _burnRedeem.burnSet[i].requiredCount, "Invalid number of tokens");
+        }
+
+        // Do redeem
+        _mint(creatorContractAddress, index, _burnRedeem, msg.sender);
+    }
+
+    /**
+     * See {ICreatorExtensionTokenURI-tokenURI}.
+     */
+    function tokenURI(address creatorContractAddress, uint256 tokenId) external override view returns(string memory uri) {
+        RedeemToken memory token = _redeemTokens[creatorContractAddress][tokenId];
+        require(token.burnRedeemIndex > 0, "Token does not exist");
+        BurnRedeem memory _burnRedeem = _burnRedeems[creatorContractAddress][token.burnRedeemIndex];
+
+        string memory prefix = "";
+        if (_burnRedeem.storageProtocol == StorageProtocol.ARWEAVE) {
+            prefix = ARWEAVE_PREFIX;
+        } else if (_burnRedeem.storageProtocol == StorageProtocol.IPFS) {
+            prefix = IPFS_PREFIX;
+        }
+        uri = string(abi.encodePacked(prefix, _burnRedeem.location));
+
+        if (!_burnRedeem.identical) {
+            uri = string(abi.encodePacked(uri, "/", uint256(token.mintNumber).toString()));
+        }
+    }
+
+    /**
+     * @dev See {IERC721Receiver-onERC721Received}.
+     */
+    function onERC721Received(
+        address,
+        address from,
+        uint256 id,
+        bytes calldata data
+    ) external override returns(bytes4) {
+        _onERC721Received(from, id, data);
+        return this.onERC721Received.selector;
+    }
+
+    /**
+     * @notice ERC721 token transfer callback
+     * @param from      the person sending the tokens
+     * @param id        the token id of the burn token
+     * @param data      bytes indicating the target burnRedeem and, optionally, a merkle proof that the token is valid
+     */
+    function _onERC721Received(
+        address from,
+        uint256 id,
+        bytes calldata data
+    ) private {
+        // Check calldata is valid
+        require(data.length % 32 == 0, "Invalid data");
+
+        address creatorContractAddress;
+        uint256 burnRedeemIndex;
+        uint256 burnItemIndex;
+        bytes32[] memory merkleProof;
+        (creatorContractAddress, burnRedeemIndex, burnItemIndex, merkleProof) = abi.decode(data, (address, uint256, uint256, bytes32[]));
```

Comment:
don't we need to encode the group index as well? which group + which item in that group we're burning for?

Reply:
also why is this `bytes32[]`, i was expecting just `bytes32`?

Reply:
also, should this just be 0'd out bytes32 for when the validationType isn't `MERKLE_TREE`?

Reply:
> also why is `merkleProof` `bytes32[]`, i was expecting just `bytes32`?

Check out the `verify` function signature here: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/MerkleProof.sol#L27

Reply:
> also, should this just be 0'd out bytes32 for when the validationType isn't `MERKLE_TREE`?

Ya see unit tests for examples with no merkle proofs